int8 loc
bool full
